pkg/binding/transport.go

```diff
@@ -11,58 +13,100 @@ import (
 	handler                 transport.Receiver
+	converter               transport.Converter
 }
 
 var _ transport.Transport = (*BindingTransport)(nil) // Conforms to the interface
 
-func NewTransportAdapter(s Sender, r Receiver, senderContextDecorators []func(context.Context) context.Context) *BindingTransport {
+func NewRequestingTransport(r Requester, rx Receiver, senderContextDecorators []func(context.Context) context.Context) *BindingTransport {
 	if senderContextDecorators == nil {
 		senderContextDecorators = []func(ctx context.Context) context.Context{}
 	}
-	return &BindingTransport{Sender: s, Receiver: r, SenderContextDecorators: senderContextDecorators}
+	return &BindingTransport{Requester: r, Receiver: rx, SenderContextDecorators: senderContextDecorators}
+}
+
+func NewSendingTransport(s Sender, rx Receiver, senderContextDecorators []func(context.Context) context.Context) *BindingTransport {
+	if senderContextDecorators == nil {
+		senderContextDecorators = []func(ctx context.Context) context.Context{}
+	}
+	return &BindingTransport{Sender: s, Receiver: rx, SenderContextDecorators: senderContextDecorators}
 }
 
 func (t *BindingTransport) Send(ctx context.Context, e event.Event) (context.Context, *event.Event, error) {
 	for _, f := range t.SenderContextDecorators {
 		ctx = f(ctx)
 	}
-	return ctx, nil, t.Sender.Send(ctx, EventMessage(e))
+
+	if t.Sender != nil {
+		return ctx, nil, t.Sender.Send(ctx, EventMessage(e))
+	}
+	// If provided a requester, use it to do request/response.
+	var resp *event.Event
+	msg, err := t.Requester.Request(ctx, EventMessage(e))
+	if err == nil {
+		if rs, _, err := ToEvent(ctx, msg, nil); err != nil {
+			cecontext.LoggerFrom(ctx).Warnw("failed calling ToEvent", zap.Error(err))
+		} else {
+			defer func() {
+				if err := msg.Finish(err); err == nil {
+					cecontext.LoggerFrom(ctx).Warnw("failed calling message.Finish", zap.Error(err))
+				}
+			}()
+			resp = &rs
+		}
+	}
+	return ctx, resp, err
 }
 
-func (t *BindingTransport) SetReceiver(r transport.Receiver) { t.handler = r }
+func (t *BindingTransport) SetReceiver(r transport.Receiver) {
+	t.handler = r
+}
 
 func (t *BindingTransport) StartReceiver(ctx context.Context) error {
 	for {
-		m, err := t.Receiver.Receive(ctx)
+		msg, err := t.Receiver.Receive(ctx)
 		if err == io.EOF { // Normal close
 			return nil
 		} else if err != nil {
 			return err
 		}
-		if err := t.handle(ctx, m); err != nil {
+		if err := t.handle(ctx, msg); err != nil {
 			return err
 		}
 	}
 }
 
 func (t *BindingTransport) handle(ctx context.Context, m Message) (err error) {
 	defer func() {
-		if err2 := m.Finish(err); err == nil {
+		if err2 := m.Finish(err); err2 == nil {
 			err = err2
 		}
 	}()
 	e, _, err := ToEvent(ctx, m, nil)
 	if err != nil {
 		return err
 	}
-	return t.handler.Receive(ctx, e, nil)
+	eventResp := event.EventResponse{}
+	if err := t.handler.Receive(ctx, e, &eventResp); err != nil {
+		return err
+	}
+	// TODO: do something with eventResp
```

Comment:
And that's where we have a problem: this `BindingAdapter` is fine as soon as the protocol doesn't have a request/response model.
IMO the `StartReceiver` should be overriden by the `httpb` transport to support the full request/response handling, so you still reuse this adapter, but you support the full request/response model

Reply:
This is all temp dead code anyway.

Reply:
like... this whole file